### PR TITLE
feat!: unified migration logic in server electron, and browser

### PR DESCRIPTION
**Pull request by himself65**
Both server, web, and electron need to consider data migration

- [x] Electron import/export workspace
  - [ ] test suite (add later because we don't release desktop version now)
- [x] Server e2e data migration test
- [x] Preloading page update
- [x] Core data upgrade e2e test from 0.8.3 stable

### CODE REVIEW

**Review by himself65 — COMMENTED** [review]
@Saul-Mirone Please help me!

**Review by Saul-Mirone — APPROVED** [review]

File: apps/electron/src/helper/dialog/dialog.ts [hunk 1]
```diff
@@ -223,6 +224,24 @@ export async function loadDBFile(): Promise<LoadDBFileResult> {
...
+    } catch (error) {
+      logger.warn(
+        `loadDBFile, migration version column failed: ${originalPath}`,
+        error
+      );
+      return { error: 'DB_FILE_MIGRATION_FAILED' };
+    } finally {
+      await db.close();
+    }
+
```
Comment (pengx17) [inline apps/electron/src/helper/dialog/dialog.ts:244]: Please move all migration related codes to `../db/migration` instead.
Also, always replace the cloned db's doc rows with the migrated ones.

File: apps/server/src/modules/sync/events/events.gateway.ts [hunk 2]
```diff
@@ -105,6 +108,58 @@ export class EventsGateway implements OnGatewayConnection, OnGatewayDisconnect {
       return false;
     }
 
+    // check migration in the rootDoc
+    if (message.workspaceId === message.guid) {
```
Comment (pengx17) [inline apps/server/src/modules/sync/events/events.gateway.ts:112]: @thorseraq please take a look

Comment (thorseraq) [reply apps/server/src/modules/sync/events/events.gateway.ts:112]: LGTM

File: apps/server/src/modules/sync/events/events.gateway.ts [hunk 3]
```diff
@@ -105,6 +108,58 @@ export class EventsGateway implements OnGatewayConnection, OnGatewayDisconnect {
...
+              .map(subdoc =>
+                this.docManager.push(
+                  message.workspaceId,
+                  subdoc.guid,
+                  Buffer.from(encodeStateAsUpdate(subdoc))
+                )
+              )
+          );
+
+          await this.docManager.apply();
```
Comment (thorseraq) [inline apps/server/src/modules/sync/events/events.gateway.ts:158]: not needed, there's an interval, it's periodically called

File: apps/server/src/modules/sync/events/events.gateway.ts [hunk 4]
```diff
@@ -105,6 +108,58 @@ export class EventsGateway implements OnGatewayConnection, OnGatewayDisconnect {
...
+          await Promise.all(
+            subdocs
+              .filter((v): v is YDoc => !!v)
+              // always upgrade the page doc to the latest version
+              .map(subdoc => schema.upgradePage(currentVersions, subdoc))
+          ).then(() =>
+            subdocs
+              .filter((v): v is YDoc => !!v)
+              .map(subdoc =>
+                this.docManager.push(
```
Comment (thorseraq) [inline apps/server/src/modules/sync/events/events.gateway.ts:150]: please only push updates, what's pushed will all be merged with snapshots later

**Comment by thorseraq** [comment]
One thing I think worth noting: If server has updated doc to latest version, and there's an old client, it will download the upgraded doc, the old client may not support rendering that

**Review by thorseraq — APPROVED** [review]

**Review by thorseraq — APPROVED** [review]

**Review by darkskygit — APPROVED** [review]

File: apps/core/src/components/migration-fallback.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    );
+    const indexedDbProvider = providers.find(
+      (provider): provider is LocalIndexedDBBackgroundProvider =>
+        provider.flavour === 'local-indexeddb-background'
+    );
+    const provider = sqliteProvider || indexedDbProvider;
+    assertExists(provider, 'no local provider');
+    return provider;
+  }, [providers]);
+  const handleClick = useCallback(async () => {
```
Comment (pengx17) [inline apps/core/src/components/migration-fallback.tsx:37]: may disable the button when migration is in progress

**Pull request merged**